File: frontend/src/types/courses.ts
export interface ModuleListItem {
  id: number;
  title: string;
  description: string | null;
  order: number;
  is_published: boolean;
  is_locked: boolean;
  lessons_count: number;
  completed_lessons_count: number;
}

export interface ModuleListResponse {
  items: ModuleListItem[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

export interface LessonSummary {
  id: number;
  title: string;
  order: number;
  estimated_duration: number | null;
  is_published: boolean;
  is_locked: boolean;
  progress_status: string | null;
}

export interface ModuleDetail {
  id: number;
  title: string;
  description: string | null;
  order: number;
  is_published: boolean;
  is_locked: boolean;
  lessons: LessonSummary[];
  created_at: string;
  updated_at: string;
}

export interface LessonDetail {
  id: number;
  module_id: number;
  title: string;
  content: string | null;
  theory: string | null;
  sql_solution: string | null;
  expected_result: Record<string, any> | null;
  order: number;
  estimated_duration: number | null;
  is_published: boolean;
  is_locked: boolean;
  progress_status: string | null;
  attempts: number;
  created_at: string;
  updated_at: string;
}

export type ModuleStatus = "locked" | "available" | "in-progress" | "completed";

export const getModuleStatus = (module: ModuleListItem): ModuleStatus => {
  if (module.is_locked) return "locked";
  if (module.completed_lessons_count === module.lessons_count && module.lessons_count > 0) {
    return "completed";
  }
  if (module.completed_lessons_count > 0) return "in-progress";
  return "available";
};

export const getStatusIcon = (status: ModuleStatus): string => {
  switch (status) {
    case "locked":
      return "🔒";
    case "available":
      return "➡️";
    case "in-progress":
      return "🟡";
    case "completed":
      return "✅";
  }
};

File: backend/app/schemas/courses.py
from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict


class LessonSummary(BaseModel):
    """Краткая информация об уроке для отображения в списке модуля."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    title: str
    order: int
    estimated_duration: int | None
    is_published: bool
    is_locked: bool
    progress_status: str | None = None


class ModuleListItem(BaseModel):
    """Элемент списка модулей."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    title: str
    description: str | None
    order: int
    is_published: bool
    is_locked: bool
    lessons_count: int
    completed_lessons_count: int


class ModuleListResponse(BaseModel):
    """Ответ со списком модулей с пагинацией."""

    items: list[ModuleListItem]
    total: int
    page: int
    page_size: int
    total_pages: int


class ModuleDetail(BaseModel):
    """Детальная информация о модуле с уроками."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    title: str
    description: str | None
    order: int
    is_published: bool
    is_locked: bool
    lessons: list[LessonSummary]
    created_at: datetime
    updated_at: datetime


class LessonDetail(BaseModel):
    """Детальная информация об уроке."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    module_id: int
    title: str
    content: str | None
    theory: str | None
    sql_solution: str | None
    expected_result: dict[str, Any] | None
    order: int
    estimated_duration: int | None
    is_published: bool
    is_locked: bool
    progress_status: str | None = None
    attempts: int = 0
    created_at: datetime
    updated_at: datetime
